### PR TITLE
refactor: PoSt Detected Faults uses a HAMT set

### PR DESCRIPTION
**Pull request by frrist**
Rebased on #54 to implement a set by wrapping a map, will need to merge it before this.

### CODE REVIEW

File: actors/builtin/storage_power/storage_power_actor_state.go [hunk 1]
```diff
@@ -191,7 +192,9 @@ func (st *StoragePowerActorState) _updatePowerEntriesFromClaimedPower(s adt.Stor
 	// Currently, the only reason for these to differ is if the miner is in DetectedFault state
 	// from a SurprisePoSt challenge.
 	nominalPower := claimedPower
-	if st.PoStDetectedFaultMiners[minerAddr] {
+	if found, err := st.hasFault(s, minerAddr); err != nil {
+		return err
+	} else if !found {
```
Comment (anorth) [inline actors/builtin/storage_power/storage_power_actor_state.go]: This looks like an inversion from the original.

File: actors/builtin/storage_power/storage_power_actor_state.go [hunk 2]
```diff
@@ -277,6 +280,36 @@ func (st *StoragePowerActorState) _getPledgeSlashForConsensusFault(currPledge ab
...
+func (st *StoragePowerActorState) hasFault(s adt.Store, a addr.Address) (bool, error) {
+	hm := adt.AsSet(s, st.PoStDetectedFaultMiners)
+	found, err := hm.Has(asKey(a))
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get detected faults for address %v from set %s", a, st.PoStDetectedFaultMiners)
+	}
+	if !found {
+		return false, nil
+	}
+	return true, nil
```
Comment (anorth) [inline actors/builtin/storage_power/storage_power_actor_state.go]: `return found, nil`.

File: actors/builtin/storage_power/storage_power_actor_state.go [hunk 3]
```diff
@@ -277,6 +280,36 @@ func (st *StoragePowerActorState) _getPledgeSlashForConsensusFault(currPledge ab
 	}
 }
 
+func (st *StoragePowerActorState) hasFault(s adt.Store, a addr.Address) (bool, error) {
+	hm := adt.AsSet(s, st.PoStDetectedFaultMiners)
```
Comment (anorth) [inline actors/builtin/storage_power/storage_power_actor_state.go]: The name `hm` isn't informative here. `faultyMiners` or similar?

File: actors/util/adt/set.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+type hamtSetValue uint64
+
+func (s hamtSetValue) MarshalCBOR(w io.Writer) error {
+	if _, err := w.Write(cbg.CborEncodeMajorType(cbg.MajUnsignedInt, uint64(s))); err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (anorth) [inline actors/util/adt/set.go]: It took me a while to figure out what's going on here, but I get it now - a sentinel value because you can't use nil.
- I think you can use `struct{}` (regardless of how we CBOR it), which will be cheaper in Go and more obviously empty, just like when used as a map value
- since it always encodes to the same byte, hardcode it rather than compute it
- is there prior art for using the value 1? We could use any single-byte encoding, such as the same thing as `EmptyReturn`, again a better representation of "nothing". 

Basically I think you can move/rename `EmptyReturn` to `adt.EmptyValue` and use it here.

File: actors/util/adt/set.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		m: AsMap(s, r),
+	}
+}
+
+// NewSet creates a new HAMT with root `r` and store `s`.
+func MakeEmptySet(s Store) (*Set, error) {
+	nd := hamt.NewNode(s)
+	newSet := AsSet(s, cid.Undef)
+	err := newSet.write(nd)
+	return newSet, err
```
Comment (anorth) [inline actors/util/adt/set.go]: Use MakeEmptyMap and wrap the result rather than duplicate the internals. Remove the then-unused `write` implementation.

File: actors/util/adt/set.go (hunk 4 above)
Comment (frrist) [reply actors/util/adt/set.go]: Prior art here: https://github.com/filecoin-project/lotus/blob/master/chain/actors/actor_storagepower.go#L751

> Basically I think you can move/rename EmptyReturn to adt.EmptyValue and use it here.

will give it a try, good idea!

File: actors/util/adt/set.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	buf := make([]byte, 1)
+	_, err := r.Read(buf)
+	if err != nil {
+		return err
+	}
+	if buf[0] != emptyListEncoded {
+		return fmt.Errorf("invalid empty return %x", buf[0])
+	}
+	return nil
+}
```
Comment (anorth) [inline actors/util/adt/set.go]: Move this to `empty.go`.
